android/trifle/src/main/java/app/cash/trifle/TrifleApi.kt

```diff
@@ -59,4 +62,37 @@ object TrifleApi {
     keyHandle: KeyHandle,
     certificates: List<Certificate>
   ): SignedData = Trifle.EndEntity(keyHandle.keyPair).createSignedData(data, certificates)
+
+  /**
+   * Verify that the provided certificate matches what we expected.
+   * It matches the CSR that we have and the root cert is what we expect.
+   *
+   * @param certificate - the certificate to verify
+   * @param certificateRequest - request used to generate this certificate
+   * @param ancestorCertificateChain - list of certificates preceding *this* one.  Namely, the first
+   *   entry should be the certificate corresponding to the issuer of this certificate, and each
+   *   thereafter should be the issuer of the one before it.
+   * @param anchorCertificate - the trust anchor against which we would like to verify the
+   *   ancestorCertificateChain. This may be the terminal (root) certificate of the chain or may be
+   *   an intermediate certificate in the chain which is already trusted.
+   * @param date - The date to use for verification against certificates' validity windows. If null,
+   *   the current time is used.
+   *
+   * @return - [Result] indicating [Result.isSuccess] or [Result.isFailure]:
+   * - success value is expressed as a [Unit] (Nothing)
+   * - failure value is expressed as a [TrifleErrors]
+   */
+  fun verify(
+    certificate: Certificate,
+    certificateRequest: CertificateRequest,
+    ancestorCertificateChain: List<Certificate>,
+    anchorCertificate: Certificate,
+    date: Date? = null
+  ): Result<Unit> {
```

Comment:
Is there a reason why we want to separate out certificate and certificate chain?
we have 
- certificate
- intermediate chain
- anchor certificate

Is it possible to collapse these 3?

Reply:
This wasn't a changed as part of the PR. The difference in behavior of using the this API deferred from current to this change is as follow:

```
deviceCert.(csr, chain, optionalAnchor, optionalDate)

vs

TrifleApi.(deviceCert, csr, chain, optionalAnchor, optionalDate)
```

Wanting to collapse them is a further departure from the initial API design. I had definitely questioned about having them collapsed, but was at least convinced that the optional anchorCertificate parameter might be something we want from past conversations where we have a different trust anchor from the certificate bundle

Reply:
I agree with Kyle - the API is very complicated. Apps should not have to know all these details.
If we look at iOS SecTrustCreateWithCertificates(), java JCACertPathValidator.validate(), they always collapse the certificates.
One of the reasons I was also pushing for moving Verify from Certificate to TrifleAPI is that it was too much to expect apps to parse the client cert, get the intermediates, anchor etc etc

I propose we go for a simple API:
`verify(cert, csr?, anchor?)`

where csr and anchor are both optional.
cert is a [TrifleCertificate]  = device client -> intermediate ... -> anchor...